File: Array/majority-element.py
#Solution 1: Using Hashmap
#Time Complexity: O(n)
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        majority_element = {}
        for i in nums:
            if i not in majority_element:
                majority_element[i] = nums.count(i)
        max_value = max(zip(majority_element.values(), majority_element.keys()))[1]
        return max_value

#Solution 2: Using Boyer Moore Algorithm
#Time Complexity: O(1)
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        res,count = 0,0
        for i in nums:
            if count == 0:
                res = i
            if i == res:
                count += 1
            elif i != res:
                count -= 1 
        return res

File: Array/Java/majorityElement.java
import java.util.HashMap;


public class majorityElement {
    // HashMap approach   
    class SolutionHashMap {
        public int majorityElement(int[] nums) {
            if(nums.length < 2){
                return nums[0];
            }
            HashMap<Integer, Integer> hashMap = new HashMap<>();
            for(int i = 0; i <= nums.length; i++){
                if(hashMap.containsKey(nums[i])){
                    hashMap.put(nums[i], hashMap.get(nums[i]) + 1);
                    if(hashMap.get(nums[i]).intValue() > nums.length / 2){
                        return nums[i];
                    }
                }
                else{
                    hashMap.put(nums[i], 1);
                }
            }
            return -1;
        }
    }
    
    //Moore Voting appraoch
    class SolutionMoore{
        public int majorityElement(int[] nums) {
            int res = nums[0];
            int count = 1;
            for(int i = 1; i < nums.length; i++){
                if(count == 0){
                    res = nums[i];
                }
                if(res == nums[i]){
                    count += 1;
                }
                else if(res != nums[i]){
                    count -= 1;
                }
            }
        return res;
        }
    }
}
